### DISCUSSION TITLE
Question About Checking Out Connection

### DISCUSSION THREAD

**Discussion by jvanderen1** [jeremyevans/sequel#2210] [General]
I was wondering if someone could clarify what [this statement](https://sequel.jeremyevans.net/rdoc-plugins/files/lib/sequel/extensions/connection_validator_rb.html#:~:text=set%20to%20%2D1%20to%20always%20validate%20connections%20on%20checkout) means:

> set to -1 to always validate connections on checkout

Specifically, what does `checkout` mean here? Does this mean Sequel will attempt to validate a connection before every query executed or something else?

We are thinking this plugin will help us resolve our `Sequel::DatabaseDisconnectError` error's, but we don't know exactly what value to add here. Our infrastructure tends to proxy connections, thus we see underlying `PG::UnableToSend` errors pretty sporadically. We think setting it to `-1` may be the best option here, but are unsure.

**Comment by jeremyevans** [comment]
Checkout means a call to Database#synchronize that results in a connection being checked out from the pool.  This happens internally whenever a connection is needed.  It's not just for queries, as a connection may be needed to correctly escape an SQL string.  If you set the value to `-1`, it may be a good idea to explicitly checkout connections around code that will use them, to reduce the number of checkouts performed internally.  Otherwise, you may see some performance degradation with the setting.

Setting it to a value such as `1` would mean only validate the connection if it has been at least 1 second since it was checked in.  If `-1` causes too much performance degradation, you could try `1` or `0.1`.

Note that if the disconnect happens while a connection is already checked out, that is not something the connection_validator extension can fix.  It can only validate connections before checkout to make sure they are working.
